File: src/main/java/io/github/takenoha/towerdefense/paper/CoreBlockTagger.java
package io.github.takenoha.towerdefense.paper;

import io.github.takenoha.towerdefense.persistence.CorePlacement;
import java.util.Objects;
import org.bukkit.block.Block;

/**
 * Checks the physical material at a database-owned core coordinate.
 *
 * <p>The current core block is not a tile entity, so it cannot carry the placement PDC that the
 * former beacon representation used.  The database placement row and {@code CoreRegistry} are
 * the source of truth; callers only invoke these checks at a coordinate from that row.</p>
 */
public final class CoreBlockTagger {
    public CoreBlockTagger(org.bukkit.plugin.Plugin plugin) {
        Objects.requireNonNull(plugin, "plugin");
    }

    public boolean tag(Block block, CorePlacement placement) {
        Objects.requireNonNull(block, "block");
        Objects.requireNonNull(placement, "placement");
        return CoreMaterialPolicy.isCurrentBlock(block.getType());
    }

    public boolean matches(Block block, CorePlacement placement) {
        Objects.requireNonNull(block, "block");
        Objects.requireNonNull(placement, "placement");
        return CoreMaterialPolicy.isCoreBlockMaterial(block.getType());
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/paper/CoreBlockTagger.kt
package io.github.takenoha.towerdefense.paper

import io.github.takenoha.towerdefense.persistence.CorePlacement
import java.util.Objects
import org.bukkit.block.Block
import org.bukkit.plugin.Plugin

/** Checks the physical material at a database-owned core coordinate. */
class CoreBlockTagger(plugin: Plugin) {
    init {
        Objects.requireNonNull(plugin, "plugin")
    }

    fun tag(block: Block, placement: CorePlacement): Boolean {
        Objects.requireNonNull(block, "block")
        Objects.requireNonNull(placement, "placement")
        return CoreMaterialPolicy.isCurrentBlock(block.type)
    }

    fun matches(block: Block, placement: CorePlacement): Boolean {
        Objects.requireNonNull(block, "block")
        Objects.requireNonNull(placement, "placement")
        return CoreMaterialPolicy.isCoreBlockMaterial(block.type)
    }
}
